adhocracy4/polls/api.py

```diff
@@ -93,57 +100,108 @@ def retrieve(self, request, *args, **kwargs):
-    def vote(self, request, pk):
-        if (
-            hasattr(settings, "A4_USE_ORGANISATION_TERMS_OF_USE")
-            and settings.A4_USE_ORGANISATION_TERMS_OF_USE
-            and not self._user_has_agreed(self.request.user)
+    def __verify_captcha(self, answer, session):
+        if hasattr(settings, "CAPTCHA_TEST_ACCEPTED_ANSWER"):
+            return answer == settings.CAPTCHA_TEST_ACCEPTED_ANSWER
+
+        if not hasattr(settings, "CAPTCHA_URL"):
+            return False
+
+        url = settings.CAPTCHA_URL
+        data = {"session_id": session, "answer_id": answer, "action": "verify"}
+        response = requests.post(url, data)
+        return json.loads(response.text)["result"]
+
+    def check_captcha(self):
+        if not self.request.data.get("captcha", False):
+            raise ValidationError(_("Your answer to the captcha was wrong."))
+
+        combined_answer = self.request.data["captcha"].split(":")
+        if len(combined_answer) != 2:
+            raise ValidationError(
+                _("Something about the answer to the captcha was wrong.")
+            )
+
+        answer, session = combined_answer
+
+        if not re.match(r"[0-9a-zA-ZäöüÄÖÜß]+", answer) or not re.match(
+            r"[0-9a-fA-F]+", session
         ):
-            if (
-                "agreed_terms_of_use" in self.request.data
-                and self.request.data["agreed_terms_of_use"]
-            ):
-                OrganisationTermsOfUse = self._get_org_terms_model()
-                OrganisationTermsOfUse.objects.update_or_create(
-                    user=self.request.user,
-                    organisation=self.get_object().project.organisation,
-                    defaults={"has_agreed": self.request.data["agreed_terms_of_use"]},
-                )
+            raise ValidationError(
+                _("Something about the answer to the captcha was wrong.")
+            )
+
+        if not self.__verify_captcha(answer, session):
+            raise ValidationError(_("Your answer to the captcha was wrong."))
+
+    def check_terms_of_use(self):
+        if getattr(
+            settings, "A4_USE_ORGANISATION_TERMS_OF_USE", False
+        ) and not self._user_has_agreed(self.request.user):
+            if self.request.data.get("agreed_terms_of_use", False):
+                if self.request.user.is_authenticated:
+                    OrganisationTermsOfUse = self._get_org_terms_model()
+                    OrganisationTermsOfUse.objects.update_or_create(
+                        user=self.request.user,
+                        organisation=self.get_object().project.organisation,
+                        defaults={
+                            "has_agreed": self.request.data["agreed_terms_of_use"]
+                        },
+                    )
             else:
                 raise ValidationError(
                     {_("Please agree to the organisation's terms of use.")}
                 )
```

Comment:
how the non-registered user will agree to the terms? This happens by default when they register IIRC